mutiny-core/src/scorer.rs

```diff
@@ -0,0 +1,374 @@
+        .map(|pubkey_str| NodeId::from_str(pubkey_str).expect("only pubkeys involved"))
+        .collect()
+}
+
+pub(crate) type ProbScorer = ProbabilisticScorer<Arc<NetworkGraph>, Arc<MutinyLogger>>;
+
+pub(crate) struct HubPreferentialScorer {
+    inner: ProbScorer,
+    preferred_hubs_set: HashSet<NodeId>,
+}
+
+impl HubPreferentialScorer {
+    pub(crate) fn new(inner: ProbScorer) -> Self {
+        Self {
+            inner,
+            preferred_hubs_set: build_preferred_hubs_set(),
+        }
+    }
+
+    fn is_preferred_hub(&self, node_id: &NodeId) -> bool {
+        self.preferred_hubs_set.contains(node_id)
+    }
+}
+
+impl Score for HubPreferentialScorer {
+    type ScoreParams = ProbabilisticScoringFeeParameters;
+
+    fn channel_penalty_msat(
+        &self,
+        short_channel_id: u64,
+        source: &NodeId,
+        target: &NodeId,
+        usage: ChannelUsage,
+        score_params: &Self::ScoreParams,
+    ) -> u64 {
+        // normal penalty from the inner scorer
+        let mut penalty =
+            self.inner
+                .channel_penalty_msat(short_channel_id, source, target, usage, score_params);
+
+        let hub_to_hub_min_penalty = (score_params.base_penalty_msat as f64 * 0.5) as u64;
+        let entering_highway_min_penalty = (score_params.base_penalty_msat as f64 * 0.7) as u64;
+
+        if self.is_preferred_hub(source) && self.is_preferred_hub(target) {
+            // Both the source and target are on the "hub highway"
+            penalty = penalty
+                .checked_mul(5)
+                .unwrap_or(u64::MAX)
+                .checked_div(10)
+                .unwrap_or(u64::MAX); // 50% discount
+            penalty = penalty
+                .checked_sub(HUB_BASE_DISCOUNT_PENALTY_MSAT)
+                .unwrap_or(hub_to_hub_min_penalty); // Base fee discount
+            penalty = penalty.max(hub_to_hub_min_penalty);
+        } else if self.is_preferred_hub(target) {
+            // Only the target is a preferred hub (entering the "hub highway")
+            penalty = penalty
+                .checked_mul(7)
+                .unwrap_or(u64::MAX)
+                .checked_div(10)
+                .unwrap_or(u64::MAX); // 30% discount
+            penalty = penalty.max(entering_highway_min_penalty);
+        } else if self.is_preferred_hub(source) {
+            // Only the source is a preferred hub (leaving the "hub highway")
+            // No discount here
+            penalty = penalty.max(score_params.base_penalty_msat);
+        } else {
+            // Neither the source nor target is a preferred hub (staying "off the highway")
+            penalty = penalty
+                .checked_mul(15)
+                .unwrap_or(u64::MAX)
+                .checked_div(10)
+                .unwrap_or(u64::MAX); // 50% extra penalty
+            penalty = penalty
+                .checked_add(HUB_BASE_DISCOUNT_PENALTY_MSAT)
+                .unwrap_or(u64::MAX); // Base fee penalty
```

Comment:
all of these `checked_X().unwrap_or(u64::MAX)` can be replaced with `saturating_X`